src/Shared/Components/CICDHistory/LogsRenderer.tsx

```diff
@@ -149,43 +171,292 @@ export const LogsRenderer = ({
+        // Would be using this to get last user action on stage
+        const previousStageMap: Readonly<Record<string, Readonly<Record<string, StageDetailType>>>> = stageList.reduce(
+            (acc, stageDetails) => {
+                if (!acc[stageDetails.stage]) {
+                    acc[stageDetails.stage] = {}
+                }
+                acc[stageDetails.stage][stageDetails.startTime] = stageDetails
+                return acc
+            },
+            {} as Record<string, Record<string, StageDetailType>>,
+        )
+
+        // Map of stage as key and value as object with key as start time and value as boolean depicting if search key is present or not
+        const searchKeyStatusMap: Record<string, Record<string, boolean>> = {}
+
+        return streamDataList.reduce((acc, streamItem: string, index) => {
+            if (streamItem.startsWith(LOGS_STAGE_IDENTIFIER)) {
+                try {
+                    const { stage, startTime, endTime, status }: StageInfoDTO = JSON.parse(streamItem.split('|')[1])
+                    const existingStage = acc.find((item) => item.stage === stage && item.startTime === startTime)
+                    const previousExistingStage = previousStageMap[stage]?.[startTime]
+
+                    if (existingStage) {
+                        // Would update the existing stage with new endTime
+                        existingStage.endTime = endTime
+                        existingStage.status = status
+                        existingStage.isOpen = getIsStageOpen(
+                            status,
+                            previousExistingStage?.isOpen,
+                            !!searchKeyStatusMap[stage]?.[startTime],
+                            !!targetSearchKey,
+                        )
+                    } else {
+                        acc.push({
+                            stage: stage || `Untitled stage ${index + 1}`,
+                            startTime: startTime || ZERO_TIME_STRING,
+                            endTime: endTime || ZERO_TIME_STRING,
+                            // Would be defining the state when we receive the end status, otherwise it is loading and would be open
+                            isOpen: true,
+                            status: StageStatusType.PROGRESSING,
+                            logs: [],
+                        })
+                    }
+                    return acc
+                } catch (e) {
+                    acc.push({
+                        stage: `Untitled stage ${index + 1}`,
+                        startTime: ZERO_TIME_STRING,
+                        endTime: ZERO_TIME_STRING,
+                        isOpen: false,
+                        logs: [],
+                        // Can not set status as FAILURE, as we will append latest logs to last stage
+                        status: StageStatusType.PROGRESSING,
+                    })
```

Comment:
these two can be commoned out

Reply:
Have removed acc.push in catch block no need of that